api/agent/drivers/docker/docker.go

```diff
@@ -345,17 +345,36 @@ func dockerMsg(derr *docker.Error) string {
 // Run executes the docker container. If task runs, drivers.RunResult will be returned. If something fails outside the task (ie: Docker), it will return error.
 // The docker driver will attempt to cast the task to a Auther. If that succeeds, private image support is available. See the Auther interface for how to implement this.
 func (drv *DockerDriver) run(ctx context.Context, container string, task drivers.ContainerTask) (drivers.WaitResult, error) {
+
+	attachSuccess := make(chan struct{})
 	mwOut, mwErr := task.Logger()
 
 	waiter, err := drv.docker.AttachToContainerNonBlocking(ctx, docker.AttachToContainerOptions{
-		Container: container, OutputStream: mwOut, ErrorStream: mwErr,
-		Stream: true, Stdout: true, Stderr: true,
-		Stdin: true, InputStream: task.Input()})
+		Success:      attachSuccess,
+		Container:    container,
+		OutputStream: mwOut,
+		ErrorStream:  mwErr,
+		Stream:       true,
+		Stdout:       true,
+		Stderr:       true,
+		Stdin:        true,
+		InputStream:  task.Input()})
+
 	if err != nil && ctx.Err() == nil {
 		// ignore if ctx has errored, rewrite status lay below
 		return nil, err
 	}
 
+	// Sync up with NB Attacher above before starting the task
+	if err == nil {
+		// WARNING: the I/O below requires docker hijack function to honor
+		// the contract below, specifically if an error is not returned
+		// from AttachToContainerNonBlocking, then max blocking time
+		// here should be what drv.docker dialer/client config was set to.
+		<-attachSuccess
```

Comment:
to be nice we should probably select here with `ctx.Done()`, for cold at least it could happen

Reply:
seems super risky, I went back and forth about this, we risk a go routine leak if we skip this...

Reply:
i see, if nobody receives this it hangs out forever. well, shit, I completely missed that part.